docs/design/README.md

```diff
@@ -534,6 +534,149 @@ After installing packages, the result of `dnf list installed` will be output to
 This will allow maintainers to audit which kit each package was taken from.
 The rest of the variant image creation steps proceed the same way that they do now.
 
+### Build Algorithm Changes
+
+Currently, the build system relies on a flat directory of RPMs.
+It relies on Cargo to build packages so that dependent RPMs are in the directory when needed.
+When creating a variant image, all RPMs are available, but the yum install command can follow package dependencies to install the right things.
+
+When creating a kit, we won't have the yum install command to help us figure out all the right packages.
+We need to change the build system so that the dependency graph can be used to put the right packages into a kit.
+
+For starters, we will no longer rely on a flat structure in `build/rpms`.
+We will put the output of each `package` build into a subdirectory using the Cargo package's name.
+
+For example, when building `glibc/Cargo.toml` we will put the output in `build/rpms/glibc`.
+Note that more than one RPM file can result from a package build.
+The subdirectory maps the resultant RPM files to the Cargo package that produced them.
```

Comment:
For knowledge sharing why is this needed to have a directory per package instead of say a directory per kit?

Reply:
> For knowledge sharing why is this needed to have a directory per package instead of say a directory per kit?

Kits may have "internal" packages that are needed to build other packages in that kit, but not intended to propagate onward when the kit is published. An example would be a package to build a host tool that is then used to build other packages for the target. To support this it's necessary to have precise targeting for which RPMs belong to which package, so that the resulting kit doesn't contain more than intended.

However, the more fundamental reason is that this is the wrong direction to think about the problem from. Packages don't "know" that they are part of a kit, any more than they should "know" they are part of a variant. `cargo` will trigger package crate builds if the kit crate depends on them, but the package doesn't know why it is being built unless we build some mechanism to pass that data in the environment. That's similar to how we pass variant data to packages today, and we're trying hard to eliminate that mechanism.